lib/uplink/main.go

```diff
@@ -0,0 +1,142 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package uplink
+
+import (
+	"context"
+
+	"storj.io/storj/pkg/transport"
+
+	"storj.io/storj/pkg/identity"
+	"storj.io/storj/pkg/metainfo/kvmetainfo"
+	"storj.io/storj/pkg/miniogw"
+	"storj.io/storj/pkg/peertls/tlsopts"
+	"storj.io/storj/pkg/storj"
+	ul "storj.io/storj/uplink"
+)
+
+// Config holds the configs for the Uplink
+type Config struct {
+	// MaxBufferMem controls upload performance and is system-specific
+	MaxBufferMem int
+
+	// These should only be relevant for new files; these values for existing
+	// files should come from the metainfo index. It's unlikely these will ever
+	// change much.
+	EncBlockSize  int
+	MaxInlineSize int
+	SegmentSize   int64
+	TLSConfig     tlsopts.Config
+}
+
+// Session represents a specific access session.
+type Session struct {
+	TransportClient transport.Client
+	Gateway         *miniogw.Gateway
+}
+
+// Access is all of the access information an application needs to store and
+// retrieve data. Someone with a share may have no restrictions within a project
+// (can create buckets, list buckets, list files, upload files, delete files,
+// etc), may be restricted to a single bucket, may be restricted to a prefix
+// within a bucket, or may even be restricted to a single file within a bucket.
+// NB(dylan): You need an Access to start a Session
+type Access struct {
+	Permissions Macaroon
```

Comment:
I attached Macaroons to the Access struct because Access structs can be per bucket, which will allow us to control permissions to a per bucket or per object resolution. This is basically another TODO here.

Reply:
Even if it's per-bucket, though, I mean, you still might need multiple macaroons. Maybe I have one for write access and a separate one for read access? Or maybe there are multiple subfolders, and I have a different macaroon for access to each

Reply:
This is why there's a session. you can make a new session per access type. that work @thepaul?

Reply:
That would make it technically possible to do all the things, but sounds like kind of a nightmare to work with for anything beyond trivially straightforward uses. Maybe I'm not envisioning the right usage scenarios.